src/Microsoft.VisualStudio.ProjectSystem.Managed/Threading/Tasks/CancellationSeries.cs

```diff
@@ -0,0 +1,112 @@
+        {
+            _superToken = token == default 
+                ? null 
+                : new[] { token };
+        }
+
+        /// <summary>
+        /// Gets the next <see cref="CancellationToken"/> in the series, ensuring the last issued
+        /// token (if any) is cancelled before this method returns.
+        /// </summary>
+        /// <param name="token">An optional cancellation token that, when cancelled, cancels the
+        /// returned token.</param>
+        /// <returns>
+        /// A cancellation token that will be cancelled when either:
+        /// <list type="bullet">
+        /// <item><see cref="GetToken"/> is called again</item>
+        /// <item>The token passed to this method (if any) is cancelled</item>
+        /// <item>The token passed to the constructor (if any) is cancelled</item>
+        /// <item><see cref="Dispose"/> is called</item>
+        /// </list>
+        /// </returns>
+        /// <exception cref="ObjectDisposedException">This object has been disposed.</exception>
+        public CancellationToken GetToken(CancellationToken token = default)
+        {
+            // Create the next CTS. There are four scenarios, depending upon whether
+            // a super token was given to the constructor, and whether a token was
+            // passed to this method.
+
+            CancellationTokenSource nextSource = _superToken == null
+                ? token == default
+                    ? new CancellationTokenSource()
+                    : CancellationTokenSource.CreateLinkedTokenSource(token)
+                : token == default
+                    ? CancellationTokenSource.CreateLinkedTokenSource(_superToken)
+                    : CancellationTokenSource.CreateLinkedTokenSource(token, _superToken[0]);
+
+            CancellationToken nextToken = nextSource.Token;
+            CancellationTokenSource priorSource = Volatile.Read(ref _cts);
+
+            // Loop until we succeed in swapping a known prior for our next.
+            // This ensures a proper sequence is observed, which is important
+            // as otherwise we may cancel a prior twice, which would coincide
+            // with a prior not being cancelled at all. We need a strict order
+            // to maintain one-in-one-out.
+            while (true)
+            {
+                if (priorSource == null)
+                {
+                    nextSource.Dispose();
+
+                    throw new ObjectDisposedException(nameof(CancellationSeries));
+                }
+
+                CancellationTokenSource snapshot = Interlocked.CompareExchange(ref _cts, nextSource, priorSource);
+
+                if (ReferenceEquals(snapshot, priorSource))
+                {
+                    break;
+                }
+
+                priorSource = snapshot;
+            }
+
+            priorSource.Cancel();
+            priorSource.Dispose();
+
+            return nextToken;
+        }
```

Comment:
This Interlocked.CompareExchange loop looks fine.  I'm just wondering if it's overkill, and if just using:
```C#
CancellationTokenSource priorSource = Interlocked.Exchange(ref _cts, nextSource);
if (priorSource == null) throw ...;
priorSource.Cancel();
priorSource.Dispose();
return nextSource.Token;
```
would be sufficient?

Reply:
I believe you're right. That's much simpler. Thank you!